File: mobile/lib/schema/schema_version.dart
/// Minimal SemVer-lite comparator (`major.minor.patch`, numeric only).
///
/// A full semver package was deliberately not added for MOBILE-RUNTIME-2:
/// this kernel only ever compares AWJ-owned schema/runtime version strings
/// of a fixed `x.y.z` shape (never pre-release/build-metadata suffixes), so
/// a narrow first-party comparator avoids a new dependency for a problem
/// this bounded (horizon MR-19: "whether a narrower first-party/framework
/// implementation is reasonable").
class SchemaVersion implements Comparable<SchemaVersion> {
  final int major;
  final int minor;
  final int patch;

  const SchemaVersion(this.major, this.minor, this.patch);

  static SchemaVersion? tryParse(String value) {
    final parts = value.trim().split('.');
    if (parts.length != 3) return null;
    final nums = <int>[];
    for (final part in parts) {
      final n = int.tryParse(part);
      if (n == null || n < 0) return null;
      nums.add(n);
    }
    return SchemaVersion(nums[0], nums[1], nums[2]);
  }

  static SchemaVersion parse(String value) {
    final version = tryParse(value);
    if (version == null) {
      throw FormatException('Invalid version string "$value" (expected "x.y.z")');
    }
    return version;
  }

  @override
  int compareTo(SchemaVersion other) {
    if (major != other.major) return major.compareTo(other.major);
    if (minor != other.minor) return minor.compareTo(other.minor);
    return patch.compareTo(other.patch);
  }

  bool operator <(SchemaVersion other) => compareTo(other) < 0;
  bool operator <=(SchemaVersion other) => compareTo(other) <= 0;
  bool operator >(SchemaVersion other) => compareTo(other) > 0;
  bool operator >=(SchemaVersion other) => compareTo(other) >= 0;

  @override
  bool operator ==(Object other) => other is SchemaVersion && compareTo(other) == 0;

  @override
  int get hashCode => Object.hash(major, minor, patch);

  @override
  String toString() => '$major.$minor.$patch';
}

File: app/Services/AppBuilder/SchemaVersion.php
<?php

namespace App\Services\AppBuilder;

/**
 * ═══════════════════════════════════════════════════════════════
 *  مقارن SemVer مصغَّر — مطابقٌ حرفياً لـ mobile/lib/schema/schema_version.dart
 * ═══════════════════════════════════════════════════════════════
 *
 * APP-BUILDER-2: «لا تخترع عقد مخطط/تشغيل ثانياً» — App Schema الحقيقي
 * المُختبَر فعلياً هو ما بنته AWJ Mobile Runtime Proof V1
 * (`mobile/lib/schema/`)، لا الشكل التوضيحي في `APP_SCHEMA_V1.md` §4 (الذي
 * يصرّح صراحةً أنه "contract candidates... not locked"). هذا الصنف نسخة
 * PHP طبق الأصل من مقارن `x.y.z` الخاص بـFlutter — رقمي فقط، بلا
 * pre-release/build-metadata، بنفس السبب: مشكلة محدودة لا تستحق حزمة SemVer
 * خارجية.
 */
final class SchemaVersion
{
    public function __construct(
        public readonly int $major,
        public readonly int $minor,
        public readonly int $patch,
    ) {}

    public static function tryParse(string $value): ?self
    {
        $parts = explode('.', trim($value));
        if (count($parts) !== 3) {
            return null;
        }

        $nums = [];
        foreach ($parts as $part) {
            if (! ctype_digit($part)) {
                return null;
            }
            $nums[] = (int) $part;
        }

        return new self($nums[0], $nums[1], $nums[2]);
    }

    public function compareTo(self $other): int
    {
        if ($this->major !== $other->major) {
            return $this->major <=> $other->major;
        }
        if ($this->minor !== $other->minor) {
            return $this->minor <=> $other->minor;
        }

        return $this->patch <=> $other->patch;
    }

    public function lessThan(self $other): bool
    {
        return $this->compareTo($other) < 0;
    }

    public function greaterThan(self $other): bool
    {
        return $this->compareTo($other) > 0;
    }

    public function __toString(): string
    {
        return "{$this->major}.{$this->minor}.{$this->patch}";
    }
}
